refactor(daser): use functional options pattern to configure daser

derrandz opened the pull request:
# Overview
The DAS module is currently reliant on default constant global variables acting as configuration variables.

This makes changing configuration values for the DAS module non-trivial, as it requires code changes and binary re-building.

In this PR, we refactor the global configuration variables used in the DAS module to use the functional options pattern, which will trivialize the process of configuring a Celestia node.

# Reference Issue
#1063 
#709 

# Impact / Change

The `das` package is now configurable using a set of functional options, namely:
```
WithSamplingRange(uint64)
WithConcurrencyLimit(int)
WithBackgroundStoreInterval(time.Duration)
WithPrioritySizeQueue(int)
WithGenesisHeight)
```

This allows the `das` package to be configured with values other than the default parameters defined in `das/options.go:29`

An example of the `das` package being configured with these options and instantiated with them is: `nodebuilder/das/module.go:22`

# Owners

@derrandz

Wondertan commented on das/daser.go:
```diff
@@ -17,28 +17,73 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+// TODO(@team): Should we DI this?
```
We use pkg level configs, so no

Wondertan commented on nodebuilder/config.go:
```diff
@@ -27,6 +28,7 @@ type Config struct {
 	RPC    rpc.Config
 	Share  share.Config
 	Header header.Config
+	DASer  das.Config
```
Thi should be a module level config as per our discussion

derrandz commented on das/coordinator_test.go:
```diff
@@ -125,13 +134,14 @@ func TestCoordinator(t *testing.T) {
 	})
 
 	t.Run("priority routine should not lock other workers", func(t *testing.T) {
+		params.samplingRange = 20
```
@walldiss tests wouldn't pass until I made this. Can you scrutinize these changes?

walldiss replied:
This line will make test pointless. The point is to have priority routine discover new headers and sample recent headers sooner than catchup workers reach it. I'll point to root of the issue in different comment

walldiss commented on das/coordinator_test.go:
```diff
@@ -72,11 +78,14 @@ func TestCoordinator(t *testing.T) {
 	})
 
 	t.Run("prioritize newly discovered over known", func(t *testing.T) {
+
+		// params := dasingParams{}
+		params.concurrencyLimit = 1
+		params.samplingRange = 4
+
```
The issue here that previously in tests variables were shadowed inside each test if they had to be changed, meaning no change of the original variable. But now you modify value of `Params` struct  inside test so it affects all tests that depend on it. In case you need to modify config, easy fix could be to make a copy of config first and then modify the copy. Looks like you had an idea of it by introducing shadowing with copy, but commented it out.
Example of solution:
```
t.Run("priority routine should not lock other workers", func(t *testing.T) {
  tParams := params
  tParams. concurrencyLimit = 100
  tParams.samplingRange = 4
  
  .. use tParams 
}
```
```

walldiss commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// Validate performs basic validation of the config.
+func (cfg *Config) Validate() error {
+	if cfg.BackgroundStoreInterval <= 0 {
+		return ErrNegativeInterval
+	}
+	return nil
+}
```
`BackgroundStoreInterval` == 0 is a valid option that suppose to disable background store I think it would be nice if it is documented in config struct. 
It would be nice to add checks for `SamplingRange`,  `ConcurrencyLimit`, `GenesisHeight` to be > 0.

derrandz replied:
These values are already `uint`s and will always be `> 0`.

walldiss replied:
They could be 0 and it is not a valid case

walldiss commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,46 @@
+
+type Config struct {
+	//  samplingRange is the maximum amount of headers processed in one job.
+	SamplingRange uint64 // TODO(@derrandz): Question to @vlad, why is this uint64?
```
it was convenient to have it as `uint64` constant, because it was compared with other `uitn54` values in code later on. For config it is better to have all values of same type (`uint`)) and convert it internally in daser where appropriate.

derrandz replied:
I agree about the point of having all types be `uint`. We should also scrutinize in the other variables that you mentioned were `uint64` and revise if they should remain so.

Tagging @Wondertan

walldiss commented on das/daser.go:
```diff
@@ -63,8 +88,10 @@ func NewDASer(
+	options ...Option,
 ) *DASer {
 	d := &DASer{
+		params:         dasingParams{},
```
Could we please start from some default config instead of zero-value here so if there some options are missing (they are optional, right?) we don't end up with 0 in some config fields. It could make DASer broken otherwise.

derrandz replied:
Exactly, I hear you. I thought about this.
Noted.

walldiss commented on das/state.go:
```diff
@@ -6,11 +6,13 @@ import (
 // coordinatorState represents the current state of sampling
 type coordinatorState struct {
-	rangeSize uint64
+	genesisHeight uint
+	samplingRange uint // is the maximum amount of headers processed in one job.
```
lets use `uint64` here to avoid multiple conversions on some operations.

derrandz replied:
I will for the time being, and will also add a `TODO` to revise whether we should keep `uint64` for the variables it's being compared with.

walldiss commented on das/state.go:
```diff
@@ -6,11 +6,13 @@ import (
-	priority   []job                      // list of headers heights that will be sampled with higher priority
-	inProgress map[int]func() workerState // keeps track of running workers
-	failed     map[uint64]int             // stores heights of failed headers with amount of attempt as value
+	priorityQueueSize uint                       // the size of the priority queue
```
store as `int` -> less convertions

walldiss commented on das/daser.go:
```diff
@@ -73,7 +100,16 @@ func NewDASer(
+	}
+
+	if len(options) == 0 {
+		log.Warn("das/daser: no options applied, daser parameters will be instantiated with zero-values, make sure to configure the daser")
```
I wonder if user could be scared by this. Like we are pushing him to do some proper set up here, but does he need to? I think it should work fine with just default config.

walldiss commented on das/coordinator_test.go:
```diff
@@ -14,18 +14,24 @@ import (
+	params.concurrencyLimit = 10
+	params.bgStoreInterval = 10 * time.Minute
+	params.priorityQueueSize = 16 * 4
+	params.genesisHeight = 1
```
nit: we already have struct literals here, so we can use it to set fields.

walldiss commented on das/coordinator.go:
```diff
@@ -9,9 +9,10 @@ import (
-	concurrencyLimit int
-	getter           header.Getter
-	sampleFn         sampleFn
+	concurrencyLimit uint
```
`int` for convenience

walldiss reviewed (changes requested):
Looks good to me. Like your functional approach with options to remove dependancy on `nodebuilder`. There are 2 things that I think could be improved:
 - types in structs could be better aligned with their usage to avoid extra conversions. 
 - we need to set up default values inside DASer in case missing options scenario.

 Besides that looks legit.

derrandz commented:
@walldiss I've addressed your comments about typing to do away with conversions in operations, however, I would still like to understand [why are the compared-with variables `uint64`s to begin with](https://github.com/celestiaorg/celestia-node/pull/1225/commits/9ac8dbdcd77811190db3c467bcf9415b386937e1#diff-b1ae42f22e1981440a95dc39c783b02b94e0f13021dc3ee21aeee926db5220c1R84)? variables like `networkHead` and so on.

renaynay commented on das/daser.go:
```diff
@@ -15,28 +15,63 @@ import (
-const (
-	//  samplingRange is the maximum amount of headers processed in one job.
-	samplingRange = 100
+type dasingParams struct {
```
```suggestion
type params struct {
```

renaynay commented on das/daser.go:
```diff
@@ -15,28 +15,63 @@ import (
 
-	// concurrencyLimit defines the maximum amount of sampling workers running in parallel.
-	concurrencyLimit = 16
+func defaultDasingParams() dasingParams {
```
```suggestion
func defaultParams() dasingParams {
```

renaynay commented on das/daser.go:
```diff
@@ -15,28 +15,63 @@ import (
-	// genesisHeight is the height sampling will start from
-	genesisHeight = 1
-)
+func WithParamConcurrencyLimit(concurrencyLimit uint) Option {
+	return func(d *DASer) {
+		d.params.concurrencyLimit = concurrencyLimit
+	}
+}
+
+func WithParamBackgroundStoreInterval(bgStoreInterval time.Duration) Option {
+	return func(d *DASer) {
+		d.params.bgStoreInterval = bgStoreInterval
+	}
+}
+
+func WithParamPriorityQueueSize(priorityQueueSize uint) Option {
+	return func(d *DASer) {
+		d.params.priorityQueueSize = priorityQueueSize
+	}
+}
+
+func WithParamGenesisHeight(genesisHeight uint) Option {
```
I don't think these options, nor `Option` belong in the `das/` pkg

This should all live at `nodebuilder/daser` since they're configuration options.

Did you already discuss the placement with @Wondertan ?

Wondertan replied:
They should be here. These are options for the DASer, not the module.

derrandz replied:
To add to @Wondertan's comment, it is with these `Option`s that the `nodebuilder/daser` will be able to "configure" the `das/daser`.

renaynay commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+var (
+	ErrNegativeInterval = errors.New("nodebuilder/daser: interval must be positive")
```
unused.

renaynay commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,44 @@
+
+// Validate performs basic validation of the config.
+func (cfg *Config) Validate() error {
+	// TODO(team): what should validate for in here? seems like all configuration fields can accept zero-values.
```
you can ensure all fields are > 0

derrandz replied:
they are `unit` so the conditional will always be truthy even if assigned `-1`

walldiss replied:
Either of `SamplingRange` == 0 (empty jobs) OR `ConcurrencyLimit` == 0 (no workers) OR `GenesisHeight ` == 0 (can't sample block height 0) will break DASer.
`BackgroundStoreInterval` == 0 disables background storer,
`PriorityQueueSize` == 0 disables prioritisation of recently produced blocks for sampling 

Lets protect user from breaking the DASer by passing 0 to config fields that could not support it.

renaynay commented on das/daser.go:
```diff
@@ -15,28 +15,63 @@ import (
 
-	// concurrencyLimit defines the maximum amount of sampling workers running in parallel.
-	concurrencyLimit = 16
+func defaultDasingParams() dasingParams {
```
also we duplicate the defaults between `nodebuilder/daser` and `das`

Wondertan replied:
The daser lib/pkg has its own defaults. The defaults in the module are specific to our node types.Ffull and light will have different defaults. 
We can make the pkg level defaults be tuned for e.g. lights, and module defaults would just use the defaults from it.

renaynay commented on das/coordinator_test.go:
```diff
@@ -13,19 +13,43 @@ import (
+	dasParams    dasingParams
+}
+
+func defaultDasParams() dasingParams {
```
why is this defined again?

derrandz replied:
good call, will remove. The original intention of this was `defaultTestDasParams` but we can just use the existing defaults

renaynay commented on das/coordinator_test.go:
```diff
@@ -231,13 +266,18 @@ func TestCoordinator(t *testing.T) {
-	samplingRange := uint64(10)
 	timeoutDelay := 5 * time.Second
 
+	params := dasingParams{}
+	params.samplingRange = 10
+	params.concurrencyLimit = 100
+	params.bgStoreInterval = 10 * time.Minute
+	params.priorityQueueSize = 16 * 4
+	params.genesisHeight = 1
```
You can just take the default and only change the fields you want different from the default params

derrandz commented:
@renaynay @walldiss @Wondertan 

Following on the thread about configuration from above, specifically this [commit](https://github.com/celestiaorg/celestia-node/pull/1225/commits/3eb8465a4881d626e70382a931f6ee3e497fd657)

I think we should think about the case **when we take the `das` pkg out of the celestia-node repository**. For the moment it makes sense to validate only at the level of the nodebuilder, but since zero values for `SamplingRange` `ConcurrencyLimit` and `GenesisHeight` do break the DASer, I was thinking about extending the functional options pattern to include errors as well:

```
type Option func (*DASer) error

func WithParamX(x uint) Option {
  return func (d *DASer) error {
      if x == 0 {
         return err
      }
      //...
  }
}
```
Such that when the constructor applies these options, it can also panic with these errors or return them and halt the DASer creation.

This is relevant because, although we start off of default parameter values in the `das` DASer, the possibility of changing these values through options opens the `das` DASer to the possibility of becoming broken due to misconfiguration.

Wondertan commented:
@derrandz, this pattern you've shown is _more_ correct independently of `das` pkg being out and we should stick to it in the daser and other pkg level options

derrandz commented:
@Wondertan Great then! I've gone ahead and did an iteration of how we can achieve innate functional options validation like discussed above. Waiting on your review.

Wondertan commented on das/daser.go:
```diff
@@ -15,28 +15,152 @@ import (
+// .       d.params.option = a
+//
+//	}
+type optionSetter func()
+type Option func(*DASer) error
+type OptionValidator func(value any, optionSetter optionSetter) error
+
+type parameters struct {
+	samplingRange     uint
+	concurrencyLimit  uint
+	bgStoreInterval   time.Duration
+	priorityQueueSize uint
+	genesisHeight     uint
+}
```
There seems to be a lot of new code for options. Let's move it into the separate file

Wondertan commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,76 @@
+	"time"
+)
+
+var (
+	ErrNoSamplingRange = errors.New("nodebuilder/daser: misconfiguration error, sampling range cannot be 0")
+	ErrNoConcLimit     = errors.New("nodebuikder/daser: misconfiguration error, concurrency limit cannot be 0")
+	ErrNoGenesisHeight = errors.New("nodebuilder/daser: misconfiguration error, genesis height cannot be 0")
+)
```
This error does not have to be exported and extracted in var, or do you see them being somehow? Because I don't

Wondertan commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,76 @@
+
+var (
+	ErrNoSamplingRange = errors.New("nodebuilder/daser: misconfiguration error, sampling range cannot be 0")
+	ErrNoConcLimit     = errors.New("nodebuikder/daser: misconfiguration error, concurrency limit cannot be 0")
```
Typo in the `builder`. But anyway, in some other PR we agreed to call our module pkg as `modp2p`, `modheader`, and `moddaser` instead of `nodebuilder/daser`.

Wondertan commented on nodebuilder/daser/config.go:
```diff
@@ -0,0 +1,76 @@
+package daser
```
Relic from the past. The pkg is called `daser` while the pkgs defining Daser is called `das`. Let's make them consistent as with any other modules we have.

Wondertan commented on das/daser.go:
```diff
@@ -15,28 +15,152 @@ import (
-	// backgroundStoreInterval is the period of time for background checkpointStore to perform a checkpoint backup.
-	backgroundStoreInterval = 10 * time.Minute
+// Option for: samplingRange
+func WithParamSamplingRange(samplingRange uint) Option {
```
Why do we need to have `Param` prefixed to each option?

derrandz replied:
Actually, we don't

Wondertan commented on das/daser.go:
```diff
@@ -15,28 +15,152 @@ import (
+// .       d.params.option = a
+//
+//	}
+type optionSetter func()
+type Option func(*DASer) error
+type OptionValidator func(value any, optionSetter optionSetter) error
```
TBH, this seems over-engineered though I totally understand where you are coming from. The approach from your [original comment](https://github.com/celestiaorg/celestia-node/pull/1225#issuecomment-1285258589) is much simpler and easier to understand. Basically, it feels to much for simple option setters

derrandz replied:
🤔 I thought about this a while before commiting to this road. It does feel over-engineered, however the reason I went with `OptionValidator` is because I did not want the `WithOption` setter to become an "un-patterned" function full of senseless if elses... But quit frankly at second thought, it is a bit too much. Reverting to the initial proposition.

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,94 @@
+
+var (
+	ErrInvalidOptionValue = func(optionName string, value string) error {
+		errorMsg := fmt.Sprintf("das/daser: invalid option value: %s, cannot be %s", optionName, value)
+		return errors.New(errorMsg)
```
Can be simplified to one-liner `fmt.Errorf`

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,94 @@
+)
+
+var (
+	ErrInvalidOptionValue = func(optionName string, value string) error {
```
Don't think this needs to be exported. It's _not_ an Err you can use to compare errors against. (The main goal of exported errors is to be comparable so that any user from the outside can check for it).

We can though defined a base error: `ErrInvalidOption = "das: invalid option"`. 
Then, we can wrap it with `fmt.Errorf("%w: value '%s' cannot be %s")`
This way, any outside user would be able to check for the base error via `errors.Is(err, das.ErrInvalidOption)` if he needs to.

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,94 @@
+
+var (
+	ErrInvalidOptionValue = func(optionName string, value string) error {
+		errorMsg := fmt.Sprintf("das/daser: invalid option value: %s, cannot be %s", optionName, value)
```
```suggestion
		errorMsg := fmt.Sprintf("das: invalid option value: %s, cannot be %s", optionName, value)
```

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	if cfg.GenesisHeight == 0 {
+		return errors.New("nodebuilder/daser: misconfiguration error, genesis height cannot be 0")
```
Missed renaming to `moddaser`

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	GenesisHeight uint
+}
+
+// TODO(@derrandz): parameters needs performance testing on real network to define optimal values
+func DefaultConfig() Config {
```
note: @derrandz, this is where `node.Type` will be passed eventually to have different Configs for each type

derrandz replied:
Noted.

derrandz replied:
Keeping track of this in #1261

Wondertan commented on das/daser.go:
```diff
@@ -71,7 +56,15 @@ func NewDASer(
+	for _, applyOpt := range options {
+		err := applyOpt(d)
+		if err != nil {
+			panic(err)
```
Let's return an error here.

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,94 @@
+}
+
+// Option for: samplingRange
+func WithSamplingRange(samplingRange uint) Option {
```
Godoc requires the name of func/var/etc in the beginning of the comment

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,94 @@
+	}
+)
+
+type Option func(*DASer) error
```
godoc

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+type Config struct {
+	//  samplingRange is the maximum amount of headers processed in one job.
+	SamplingRange uint
+
+	// concurrencyLimit defines the maximum amount of sampling workers running in parallel.
+	ConcurrencyLimit uint
+
+	// backgroundStoreInterval is the period of time for background checkpointStore to perform a checkpoint backup.
+	BackgroundStoreInterval time.Duration
+
+	// priorityQueueSize defines the size limit of the priority queue
+	PriorityQueueSize uint
+
+	// genesisHeight is the height sampling will start from
+	GenesisHeight uint
```
Capitalize comments as the names

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	if cfg.ConcurrencyLimit == 0 {
+		return errors.New("moddaser: misconfiguration error, concurrency limit cannot be 0")
```
```suggestion
		return errors.New("moddas: misconfiguration: concurrency limit cannot be 0")
```

derrandz replied:
I think `return errors.New("moddas misconfiguration: concurrency limit cannot be 0")` is better, wdyt?

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+//
+// Both of which won't break the DASer
+func (cfg *Config) Validate() error {
+	// TODO(team): what should validate for in here? seems like all cfg.nfiguration fields can accept zero-values.
```
@walldiss

derrandz replied:
He answered previously, so resolving.

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	"time"
+)
+
+type Config struct {
```
godoc

derrandz commented on nodebuilder/das/module.go:
```diff
@@ -11,11 +11,31 @@ import (
 			"daser",
+			baseComponents,
 			fx.Provide(fx.Annotate(
 				NewDASer,
```
@renaynay I want to double-check, we changed the signature for this constructor to return an error, if this errors out, will fx halt the instantion and `app.Err()` should return the error causing the halt, is this correct?

Wondertan replied:
I am not Rene, but def yes

Wondertan reviewed (commented):
Last nits. Also, we should stick to some integer types everywhere to avoid  type conversion in the params and Daser itself

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,113 @@
+
+var (
+	ErrInvalidOption      = fmt.Errorf("das: invalid option")
+	ErrInvalidOptionValue = func(optionName string, value string) error {
```
Specifically, the func. It also doesn't have to be a var and can be declared as a simple func

derrandz replied:
We should keep the `ErrInvalidOption` exported to allow for error comparison like mentioned in our previous conversation in https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r1001825643

Wondertan replied:
Yes, but I specifically mention a func

derrandz replied:
Agreed

derrandz commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+// Both of which won't break the DASer
+func (cfg *Config) Validate() error {
+	if cfg.SamplingRange == 0 {
+		return errors.New("moddas misconfiguration: sampling range cannot be 0")
```
https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r1001975784

Wondertan replied:
This is more about the error styling being inconsistent. The GenesisHeight error one

derrandz commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	if cfg.ConcurrencyLimit == 0 {
+		return errors.New("moddas misconfiguration: concurrency limit cannot be 0")
```
Same here, I think the format we agreed to is more sensible (refer to https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r1001975784)

derrandz commented on das/coordinator_test.go:
```diff
@@ -13,19 +13,31 @@ import (
-	networkHead := uint64(500)
-	sampleFrom := uint64(genesisHeight)
-	timeoutDelay := 125 * time.Second
+type testParams struct {
```
Cool 👍🏻 I like it

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	if cfg.GenesisHeight == 0 {
+		return errors.New("moddas: misconfiguration: genesis height cannot be 0")
```
Are we fixing this or not?

derrandz replied:
🤣 we are!

Wondertan commented on das/state.go:
```diff
@@ -77,17 +81,18 @@ func (s *coordinatorState) updateHead(last uint64) bool {
 	}
 
-	if s.networkHead == genesisHeight {
+	// TODO(@derrandz): question to @walldiss, why is networkHead uint64 to begin with?
```
TODO to be resolved

walldiss replied:
It is `uint64` so it is possible to compare it with other variables of same type without conversion.

derrandz replied:
@walldiss I am trying to understand which variable is the one that has to be `unit64` and that caused the rest to become `uint64` to reduce conversion.

Previously in this PR, you mentioned https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r998376041 and https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r998406090 which have lead me to believe that `networkHead` was the original variable that had to be `uint64` (_for some reason_) and thus caused the rest to become so to avoid conversion.

If there is no particular reason to using `uint64` specifically, let's revert to `uint`.

walldiss replied:
Thare are many of them, `uint64` is used for any block heights variables inside DASer.  `uint64` is used over `uint` to avoid usage of uint32 on 32bit platforms.

derrandz replied:
As per our conversation, I will revert everything in daser to uint with the argument that:

No concrete argument to raise precision to uint64 or other.
We only know so far that we need positive values for the concerned variables.
:+1:

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,114 @@
+// Refer to WithSamplingRange documentation to see an example of how to use this
+func WithConcurrencyLimit(concurrencyLimit int) Option {
+	return func(d *DASer) error {
+		if concurrencyLimit == 0 {
```
This check does not check "-" values, while `int` allows them. Also true for other options.

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,114 @@
+		if concurrencyLimit == 0 {
+			return errInvalidOptionValue(
+				"concurrencyLimit",
+				"0",
```
This should take an actual `concurrencyLimit` that could be a -

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+	SamplingRange uint
+
+	// ConcurrencyLimit defines the maximum amount of sampling workers running in parallel.
+	ConcurrencyLimit int
+
+	// BackgroundStoreInterval is the period of time for background checkpointStore to perform a checkpoint backup.
+	BackgroundStoreInterval time.Duration
+
+	// PriorityQueueSize defines the size limit of the priority queue
+	PriorityQueueSize int
```
What about making these a units?

derrandz replied:
For these, since they are used mostly in comparisons against slices lens (_`len(slice) > concurrencyLimit`, similarly for priorityQueueSize_) it makes more sense to keep them as ints to avoid conversions.

renaynay commented on das/checkpoint.go:
```diff
@@ -5,18 +5,18 @@ import (
 type checkpoint struct {
-	SampleFrom  uint64 `json:"sample_from"`
-	NetworkHead uint64 `json:"network_head"`
+	SampleFrom  uint `json:"sample_from"`
```
We don't see the added benefit of this change - it really only bloats the diff for no improvement. Also @walldiss made a point that if we decide to use atomics later, we'd need to revert anyway. Can you please revert this and use `uint64` in the config

renaynay replied:
For clarity - we're referring to all `uint64` -> `uint` changes

derrandz replied:
Claro 👍🏻

renaynay commented on das/options.go:
```diff
@@ -0,0 +1,114 @@
+	}
+}
+
+// WithBackgroundStoreInterval is a functional option to configure the daser's `backgroundStoreINterval` parameter
```
```suggestion
// WithBackgroundStoreInterval is a functional option to configure the daser's `backgroundStoreInterval` parameter
```

renaynay commented on nodebuilder/das/module.go:
```diff
@@ -11,11 +11,31 @@ import (
 
-func ConstructModule(tp node.Type) fx.Option {
+func ConstructModule(tp node.Type, cfg *Config) fx.Option {
+
```
```suggestion
```

renaynay commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+// Config contains configuration parameters for the DASer (or DASing process)
+type Config struct {
```
@derrandz can you please export `params` and its fields from `das` package and make a 
`type Config das.Params`  ? Also, we don't technically need this `DefaultConfig()` since the `daser` is taking options rather than a config / config values.

renaynay replied:
This is with the goal of deduplication of the code
For clarification: we need to keep the optional parameters approach.

derrandz replied:
Yeah I see what you are saying. This was an attempt to be compliant with the pattern used with the other modules but it's very redundant.

renaynay commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,70 @@
+// PriorityQueueSize == 0 disables prioritization of recently produced blocks for sampling
+//
+// Both of which won't break the DASer
+func (cfg *Config) Validate() error {
```
We also don't need validation since we won't provide a Config here, it would just point to das.Params

derrandz replied:
🤔 hmmm, I see what you are saying, and I agree, but for the record, if we wanted to be very strict with our patterns, here are two arguments of why you would keep this redundant layer of validation:

1. Validation in this level concerns `nodebuilder`, meaning that this is validation for values coming from a file or CLI, irrespective of where it will go (_i.e: the underlying implementation of `das` that will receive this configuration_), so we must keep it to respect the pattern. To drive this point home, we could use a different daser down the line (_we won't but you see my point_)

2. The validation of `das.paramaeters` in the functional options inside of the `das` package was made in an effort to further isolate the package from `nodebuilder` and make it more of a standalone (_per my understand, that's the purpose we are doing this, correct me if I am wrong_) thus, that validation is useful for any code that will invoke those functional parameters. For the case of nodebuilder, it just so happens that it already validates its own configuration, but for other users (_when we separate das into its own go package_) it might not be the case.

derrandz replied:
After some thought, we need to keep validation at this level, but change it slightly.

The reason is, we want `das.Config` to reflect in `das.Parameters` of the daser when `das.Config` has non-zero values retrieved from CLI or file or API, this is due to [this piece of code](https://github.com/celestiaorg/celestia-node/pull/1225/files#diff-9eef3a4c09b5618e9b60c229170908fc389f8a28ce84e29e47c82d33b74e32b2R23-R28) that retrieves configuration values and "passes them as options".

However this will cause us to change the [functional options validation ](https://github.com/celestiaorg/celestia-node/pull/1225/files#diff-fa4e1a1d6adb3bac7ebcb4524081bfce3ae7764b2c0a249c8a5655cdcac5b3bdR55-R64) to not return errors and cause the constructor to fail but rather to use default options when validation fails, and log a warning line.

I will hold until we reach a decision on this.

@renaynay @Wondertan

derrandz replied:
I've settled on:

1. Deduping the code by using `type Config das.Parameters`
2. Keeping the `(*Config) Validate` to avoid the impact of removing it that is described in [the second comment](https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r1004365160) based on argumentation from [the first comment](https://github.com/celestiaorg/celestia-node/pull/1225#discussion_r1004333336)

Wondertan replied:
@derrandz, sorry for the confusion. We didn't want to avoid the second validation, but only dedup the code and keep it grouped where it logically should be in `das`. So the point where you settled is exactly what we discussed.

derrandz replied:
Clear, I will do one more iteration on this.

derrandz replied:
@renaynay @Wondertan I think this is brilliant, thanks for the comments

renaynay commented on das/options.go:
```diff
@@ -0,0 +1,114 @@
+	concurrencyLimit  int
+	bgStoreInterval   time.Duration
+	priorityQueueSize int
+	genesisHeight     uint
```
let's rename this to `sampleFrom` since `genesis` typically always refers to 1

renaynay commented on das/state.go:
```diff
@@ -6,32 +6,36 @@ import (
 // coordinatorState represents the current state of sampling
 type coordinatorState struct {
-	rangeSize uint64
+	genesisHeight uint
```
can we rename this to `sampleFrom`

renaynay commented on das/state.go:
```diff
@@ -110,7 +114,8 @@ func (s *coordinatorState) nextJob() (next job, found bool) {
 	j := s.newJob(s.next, s.networkHead)
 
-	s.next += s.rangeSize
+	// TODO(@derrandz): same here @walldiss why is next a uint?
```
If resolved, pls remove comment.

renaynay commented on das/daser_test.go:
```diff
@@ -39,7 +39,8 @@ func TestDASerLifecycle(t *testing.T) {
 	t.Cleanup(cancel)
 
-	daser := NewDASer(avail, sub, mockGet, ds, mockService)
+	daser, dErr := NewDASer(avail, sub, mockGet, ds, mockService)
```
nit but please just use `err`

renaynay commented on das/daser.go:
```diff
@@ -156,6 +149,7 @@ func (d *DASer) sample(ctx context.Context, h *header.ExtendedHeader) error {
 		var byzantineErr *share.ErrByzantine
 		if errors.As(err, &byzantineErr) {
 			log.Warn("Propagating proof...")
+			// TODO(team): remove type conversion uint64(h.Height) in dedicated PR that deals with the fraud service
```
We cannot do this, so it can be removed

renaynay commented on das/daser.go:
```diff
@@ -71,9 +56,17 @@ func NewDASer(
+	}
+
+	d.sampler = newSamplingCoordinator(d.params, getter, d.sample)
+
```
```suggestion
```

renaynay reviewed (changes requested):
Reviewed by @Wondertan @distractedm1nd @walldiss @vgonkivs

Wondertan commented on das/coordinator.go:
```diff
@@ -9,9 +9,10 @@ import (
-	concurrencyLimit int
-	getter           header.Getter
-	sampleFn         sampleFn
+	ConcurrencyLimit int
```
Why exported?

Wondertan commented on das/coordinator.go:
```diff
@@ -128,7 +129,7 @@ func (sc *samplingCoordinator) getCheckpoint(ctx context.Context) (checkpoint, e
-// concurrencyLimitReached indicates whether concurrencyLimit has been reached
-func (sc *samplingCoordinator) concurrencyLimitReached() bool {
-	return len(sc.state.inProgress) >= sc.concurrencyLimit
+// ConcurrencyLimitReached indicates whether ConcurrencyLimit has been reached
+func (sc *samplingCoordinator) ConcurrencyLimitReached() bool {
+	return len(sc.state.inProgress) >= sc.ConcurrencyLimit
```
Why exported? It seems like a relic from IDE find-and-replace

derrandz replied:
Indeed

Wondertan commented on das/state.go:
```diff
@@ -6,11 +6,13 @@ import (
 // coordinatorState represents the current state of sampling
 type coordinatorState struct {
-	rangeSize uint64
+	SampleFrom    uint64
```
Again

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,50 @@
+//
+// Both of which won't break the DASer
+func (cfg *Config) Validate() error {
+	if !(cfg.SamplingRange > 0) {
+		return errors.New("moddas misconfiguration: sampling range cannot be negative or 0")
+	}
+
+	if !(cfg.ConcurrencyLimit > 0) {
+		return errors.New("moddas misconfiguration: concurrency limit cannot be negative or 0")
+	}
+
+	if !(cfg.SampleFrom > 0) {
+		return errors.New("moddas misconfiguration: genesis height cannot be negative or 0")
+	}
```
___optional___ nit: `cfg.SamplingRange <= 0` is a bit more readable and as per @distractedm1nd using negation should be avoided where possible.

derrandz replied:
+1 for the nit @distractedm1nd

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,50 @@
+//
+// Both of which won't break the DASer
+func (cfg *Config) Validate() error {
+	if !(cfg.SamplingRange > 0) {
+		return errors.New("moddas misconfiguration: sampling range cannot be negative or 0")
+	}
+
+	if !(cfg.ConcurrencyLimit > 0) {
+		return errors.New("moddas misconfiguration: concurrency limit cannot be negative or 0")
+	}
+
+	if !(cfg.SampleFrom > 0) {
+		return errors.New("moddas misconfiguration: genesis height cannot be negative or 0")
+	}
```
Move to Params as a method. See https://github.com/celestiaorg/celestia-node/pull/1225/files#r1005327363

Wondertan commented on das/coordinator.go:
```diff
@@ -128,7 +129,7 @@ func (sc *samplingCoordinator) getCheckpoint(ctx context.Context) (checkpoint, e
 }
 
-// concurrencyLimitReached indicates whether concurrencyLimit has been reached
+// ConcurrencyLimitReached indicates whether ConcurrencyLimit has been reached
```
pls

derrandz replied:
thanks 👍🏻

Wondertan commented on das/options.go:
```diff
@@ -0,0 +1,142 @@
+
+// Validate validates the valuea in Parameters
+func (p *Parameters) Validate() error {
+	if p.SamplingRange <= 0 {
+		return errors.New("sampling range cannot be negative or 0")
+	}
+
+	if p.ConcurrencyLimit <= 0 {
+		return errors.New("concurrency limit cannot be negative or 0")
+	}
+
+	if p.SampleFrom <= 0 {
+		return errors.New("genesis height cannot be negative or 0")
+	}
```
Let's use the errInvalidOption here

Wondertan commented on das/daser.go:
```diff
@@ -71,9 +56,16 @@ func NewDASer(
 
-	return d
+	for _, applyOpt := range options {
+		err := applyOpt(d)
```
To avoid duplication of errors in options and Validate on Params, let's remove errors from Options and call here `Validate` after applying all the options. This way we use Validate twice on the `moddaser` and inside the Daser constructor here, but we avoid duplication which good. Suggested by @walldiss and I think this is good idea and hope that's the last fix before we can say our configuration pattern is complete

renaynay commented on das/daser_test.go:
```diff
@@ -73,7 +74,8 @@ func TestDASer_Restart(t *testing.T) {
 	t.Cleanup(cancel)
 
-	daser := NewDASer(avail, sub, mockGet, ds, mockService)
+	daser, dErr := NewDASer(avail, sub, mockGet, ds, mockService)
```
nits but everywhere should not be `dErr` but `err`.

renaynay commented on das/options.go:
```diff
@@ -0,0 +1,142 @@
+	"time"
+)
+
+var (
+	ErrInvalidOption = fmt.Errorf("das: invalid option")
+)
+
+func errInvalidOptionValue(optionName string, value string) error {
+	return fmt.Errorf("%w: value %s cannot be %s", ErrInvalidOption, optionName, value)
+}
```
We can remove this once we remove err checking in options

derrandz replied:
It will still be used in `das.Parameters.Validate()`

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,45 @@
+//
+// Both of which won't break the DASer
+func (cfg *Config) Validate() error {
+	params := das.Parameters(*cfg)
+	err := params.Validate()
+
+	if err != nil {
```
```suggestion
	err := (*das.Parameters)(cfg).Validate()
	if err != nil {
```

Wondertan replied:
A small nit/trick/optimization

renaynay commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,45 @@
+	return Config(das.DefaultParameters())
+}
+
+// Validate performs basic validation of the config.
+//
+// SamplingRange == 0 will cause the jobs' queue to be empty
+//
+//	therefore no sampling jobs will be reserved and more importantly
+//	 the DASer will break
+//
+// ConcurrencyLimit == 0 will cause the number of workers to be 0 and thus
+//
+//	no threads will be assigned to the waiting jobs therefore breaking the DASer
+//
+// OR GenesisHeight == 0 there is no block height 0 to be sampled thus will break the DASer.
+//
+// # On the other hand
+//
+// BackgroundStoreInterval == 0 disables background storer,
+// PriorityQueueSize == 0 disables prioritization of recently produced blocks for sampling
+//
+// Both of which won't break the DASer
```
Can we move this comment to the actual implementation of `Validate` on params? 
Also all of the explanations for the checks can be comments in the code itself.

renaynay commented on das/options.go:
```diff
@@ -0,0 +1,142 @@
+	}
+}
+
+// Validate validates the valuea in Parameters
```
```suggestion
// Validate validates the values in Parameters
```

also refer to my comment on das.Config pls 🙏🏻

renaynay commented:
Also @derrandz can you check all your godocs on the methods you've implemented and make it according to godoc style? 

for example: 

```
// Type Parameters .....
type Parameters struct {
```

should be 

```
// Parameters .....
type Parameters struct {
```

derrandz commented on nodebuilder/das/daser.go:
```diff
@@ -15,6 +15,7 @@ func NewDASer(
+	options ...das.Option,
+) (*das.DASer, error) {
+	return das.NewDASer(da, hsub, store, batching, fraudService, options...)
 }
```
If this file is removed, then using `das.DASer` in the [module construction](https://github.com/celestiaorg/celestia-node/pull/1225/files#diff-9eef3a4c09b5618e9b60c229170908fc389f8a28ce84e29e47c82d33b74e32b2L20) breaks DI.

The reason whyis: _Quoting @distractedm1nd_

> its because of the casting

> ```
> store header.Store,
>	batching datastore.Batching,
>	fraudService fraud.Module,
> ```
>```
>  getter header.Getter,
>	dstore datastore.Datastore,
>	bcast fraud.Broadcaster,
>```
> we had a discussion about this before (whether to keep such proxy constructors for casting, or to change the original constructor to use what the DI provides) and we decided that its better to keep this ugliness on the DI so that the constructor can take exactly what is needed

Thus we are keeping this file.

Wondertan replied:
Ok, thanks for noting that!

Wondertan commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,36 @@
+// Upon encountering an invalid value, Validate returns an error of type: ErrMisConfig
+func (cfg *Config) Validate() error {
+	err := (*das.Parameters)(cfg).Validate()
+
```
```suggestion
```

renaynay replied:
pls @derrandz

Wondertan reviewed (approved):
Thanks for bringing this over the finish line. This PR was a great way for us to develop the configuration pattern for our codebase and get everyone onboarded with it.

renaynay commented on das/daser_test.go:
```diff
@@ -146,7 +150,9 @@ func TestDASer_stopsAfter_BEFP(t *testing.T) {
 
 	// create and start DASer
-	daser := NewDASer(avail, sub, mockGet, ds, f)
+	daser, dErr := NewDASer(avail, sub, mockGet, ds, f)
```
dErr 😭

derrandz replied:
ouch, sorry! thought I finished with those

renaynay commented on nodebuilder/das/config.go:
```diff
@@ -0,0 +1,36 @@
+	"github.com/celestiaorg/celestia-node/das"
+)
+
+// ErrMisConfig is an error that is returned by (*Config).Validate
+// when supplied with invalid values.
+var ErrMisConfig = fmt.Errorf("moddas misconfiguration")
```
unnecessary error definition as it's returned in one place.

Wondertan replied:
I think it's a good idea though

derrandz replied:
yup went with rene's suggestion. At second thought, I should have stuck to this style, because it would allow `nodebuilder` to check for `das.ErrMisConfig` (_`nodebuilder/das` that is_) which is more contextualized and clean than checking for `das.ErrInvalidOption`.

I think we can address this in a separate issue when we make sure we truly want that all proxy modules under `nodebuilder` become full proxies and leak nothing about their external origins.

derrandz replied:
Tracking this in https://github.com/celestiaorg/celestia-node/issues/1289 
Can be closed if irrelevant

walldiss reviewed (approved):
Looks amazing, Great work!
